File: ai_fleet/services/gateway/routes/businesses.py
"""Businesses routes (port of services/gateway/src/routes/businesses.js).

A "business" is a local mapping of a name + repository + optional linked Linear
project. Creating one can optionally spin up a brand-new Linear project (with the
configured enrich labels auto-attached so the enrichment scheduler picks it up).
The canonical repository namespace is stored beside its explicit provider so a
later global connector switch cannot reinterpret the reference.
"""

from __future__ import annotations

import re
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ai_fleet import store, linear
from ai_fleet.services.common import json_body
from ai_fleet.agent.workspace import repo_parts

router = APIRouter()

# Sentinel distinguishing "key absent" (JS undefined → use fallback) from an
# explicit null/empty value supplied by the client.
_UNDEFINED = object()


def _iso_now():
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def slugify(name):
    slug = str(name).lower().strip()
    slug = re.sub(r"[^a-z0-9]+", "-", slug)
    slug = re.sub(r"^-+|-+$", "", slug)
    return slug or f"biz-{str(uuid.uuid4())[:8]}"


def normalize_repo(value, provider="github"):
    """Normalize an operator-supplied repository reference to its canonical namespace.

    Explicit provider identity is required so a bare namespace cannot change meaning
    when the global repository connector changes. GitLab may use nested groups.
    """
    raw = str(value or "").strip()
    if not raw:
        return ""
    parts = repo_parts(raw, provider)
    return parts["fullName"] if parts else ""


def normalize_repo_provider(value=_UNDEFINED, fallback="github"):
    raw = fallback if value is _UNDEFINED else value
    provider = str(raw).strip().lower()
    return provider if provider in ("github", "gitlab") else ""


def repository_fields(body, current=None):
    fallback_provider = current["repoProvider"] if current and current.get("repoProvider") else "github"
    repo_provider = normalize_repo_provider(body.get("repoProvider", _UNDEFINED), fallback_provider)
    if not repo_provider:
        return {"error": "Repository provider must be GitHub or GitLab."}

    if "repo" in body:
        raw_repo = str(body.get("repo") or "").strip()
    else:
        raw_repo = str((current.get("repo") if current else None) or "").strip()
    repo = normalize_repo(raw_repo, repo_provider)
    if raw_repo and not repo:
        format = (
            "group/project (nested groups are supported)"
            if repo_provider == "gitlab"
            else "owner/repository"
        )
        provider_label = "GitLab" if repo_provider == "gitlab" else "GitHub"
        return {"error": f"Repository must be a {provider_label} {format} or matching official-host Git URL."}
    return {"repo": repo, "repoProvider": repo_provider}


async def with_projects(businesses):
    """Attach the linked Linear project (name/state) to each business, if any."""
    linked_ids = [b["projectId"] for b in businesses if b.get("projectId")]
    if len(linked_ids) == 0:
        return [{**b, "project": None} for b in businesses]
    try:
        projects = await linear.get_projects(store.get_api_key())
    except Exception:
        # Without a valid key we still return businesses, just without project detail.
        return [{**b, "project": None} for b in businesses]
    by_id = {p["id"]: p for p in projects}
    return [{**b, "project": (by_id.get(b["projectId"]) if b.get("projectId") else None)} for b in businesses]


# GET /api/businesses — businesses with resolved Linear project detail.
@router.get("")
@router.get("/")
async def list_businesses():
    businesses = store.read_store()["businesses"]
    return {"businesses": await with_projects(businesses)}


# POST /api/businesses — create a business, optionally creating its Linear project.
@router.post("")
@router.post("/")
async def create_business(request: Request):
    body = await json_body(request)
    name = (str(body["name"]) if body.get("name") else "").strip()
    if not name:
        return JSONResponse(status_code=400, content={"error": "Business name is required."})

    description = str(body["description"]) if body.get("description") else ""
    repository = repository_fields(body)
    if repository.get("error"):
        return JSONResponse(status_code=400, content={"error": repository["error"]})
    project_id = str(body["projectId"]) if body.get("projectId") else None

    # Optionally create a brand-new Linear project for this business.
    if not project_id and body.get("createNewProject"):
        team_id = str(body["teamId"]) if body.get("teamId") else ""
        if not team_id:
            return JSONResponse(status_code=400, content={"error": "A team is required to create a new project."})
        # Auto-attach the configured enrich labels (default ["AI"]) so the new
        # project is immediately picked up by the enrichment scheduler. Gated by the
        # autoLabelNewProjects config toggle.
        config = store.get_agent_config()
        label_ids = []
        if config.get("autoLabelNewProjects"):
            labels = await linear.get_or_create_project_labels(store.get_api_key(), config.get("enrichLabels"))
            label_ids = [l["id"] for l in labels]
        project = await linear.create_project(
            store.get_api_key(),
            name=str(body["projectName"]) if body.get("projectName") else name,
            description=description,
            team_id=team_id,
            label_ids=label_ids,
        )
        project_id = project["id"]

    current = store.read_store()
    business = {
        "id": slugify(name),
        "name": name,
        "description": description,
        "projectId": project_id,
        # Provider is stored beside the canonical namespace so a later global
        # connector switch cannot reinterpret this project repository.
        "repo": repository["repo"],
        "repoProvider": repository["repoProvider"],
        "createdAt": _iso_now(),
    }
    if any(b.get("id") == business["id"] for b in current["businesses"]):
        return JSONResponse(status_code=409, content={"error": "A business with that name already exists."})

    next_store = {**current, "businesses": [*current["businesses"], business]}
    store.write_store(next_store)
    enriched = (await with_projects([business]))[0]
    return JSONResponse(status_code=201, content={"business": enriched})


# PUT /api/businesses/:id — update fields / link a Linear project.
@router.put("/{id}")
async def update_business(id: str, request: Request):
    current_store = store.read_store()
    index = next((i for i, b in enumerate(current_store["businesses"]) if b.get("id") == id), -1)
    if index == -1:
        return JSONResponse(status_code=404, content={"error": "Business not found."})

    body = await json_body(request)
    current = current_store["businesses"][index]
    repository = repository_fields(body, current)
    if repository.get("error"):
        return JSONResponse(status_code=400, content={"error": repository["error"]})
    updated = {
        **current,
        "name": (str(body["name"]).strip() or current["name"]) if "name" in body else current["name"],
        "description": str(body["description"]) if "description" in body else current["description"],
        "projectId": (
            (str(body["projectId"]) if body.get("projectId") else None)
            if "projectId" in body
            else current.get("projectId")
        ),
        "repo": repository["repo"],
        "repoProvider": repository["repoProvider"],
    }

    businesses = [updated if i == index else b for i, b in enumerate(current_store["businesses"])]
    store.write_store({**current_store, "businesses": businesses})
    enriched = (await with_projects([updated]))[0]
    return {"business": enriched}


# DELETE /api/businesses/:id — remove the local business mapping (leaves Linear untouched).
@router.delete("/{id}")
async def delete_business(id: str):
    current_store = store.read_store()
    businesses = [b for b in current_store["businesses"] if b.get("id") != id]
    if len(businesses) == len(current_store["businesses"]):
        return JSONResponse(status_code=404, content={"error": "Business not found."})
    store.write_store({**current_store, "businesses": businesses})
    return {"ok": True}

File: services/gateway/src/routes/businesses.js
'use strict';

const express = require('express');
const crypto = require('crypto');
const { readStore, writeStore, getApiKey, getAgentConfig } = require('@ai-fleet/shared/store');
const { getProjects, createProject, getOrCreateProjectLabels } = require('@ai-fleet/shared/linear');
const { asyncHandler } = require('@ai-fleet/shared/util');
const { repoParts } = require('@ai-fleet/shared/agent/workspace');
const { normalizeEventContext, matchesEventContext } = require('@ai-fleet/shared/messaging/events');

const router = express.Router();

function workspaceContext(req) {
  return normalizeEventContext(req && req.fleetContext ? req.fleetContext : {});
}

function inWorkspace(req, record) {
  const context = workspaceContext(req);
  return !context.organizationId || matchesEventContext(record, context);
}

function contextFields(req) {
  const context = workspaceContext(req);
  return {
    ...(context.organizationId ? { orgId: context.organizationId } : {}),
    ...(context.projectId ? { nativeProjectId: context.projectId } : {}),
  };
}

function slugify(name) {
  return String(name)
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '') || `biz-${crypto.randomUUID().slice(0, 8)}`;
}

/**
 * Normalize an operator-supplied repository reference to its canonical namespace.
 * Explicit provider identity is required so a bare namespace cannot change meaning
 * when the global repository connector changes. GitLab may use nested groups.
 */
function normalizeRepo(value, provider = 'github') {
  const raw = String(value || '').trim();
  if (!raw) return '';
  const parts = repoParts(raw, provider);
  return parts ? parts.fullName : '';
}

function normalizeRepoProvider(value, fallback = 'github') {
  const provider = String(value === undefined ? fallback : value).trim().toLowerCase();
  return provider === 'github' || provider === 'gitlab' ? provider : '';
}

function repositoryFields(body, current = null) {
  const fallbackProvider = current && current.repoProvider ? current.repoProvider : 'github';
  const repoProvider = normalizeRepoProvider(body.repoProvider, fallbackProvider);
  if (!repoProvider) return { error: 'Repository provider must be GitHub or GitLab.' };

  const rawRepo = body.repo !== undefined ? String(body.repo || '').trim() : String((current && current.repo) || '').trim();
  const repo = normalizeRepo(rawRepo, repoProvider);
  if (rawRepo && !repo) {
    const format = repoProvider === 'gitlab' ? 'group/project (nested groups are supported)' : 'owner/repository';
    return { error: `Repository must be a ${repoProvider === 'gitlab' ? 'GitLab' : 'GitHub'} ${format} or matching official-host Git URL.` };
  }
  return { repo, repoProvider };
}

/** Attach the linked Linear project (name/state) to each business, if any. */
async function withProjects(businesses) {
  const linkedIds = businesses.map((b) => b.projectId).filter(Boolean);
  if (linkedIds.length === 0) {
    return businesses.map((b) => ({ ...b, project: null }));
  }
  let projects = [];
  try {
    projects = await getProjects(getApiKey());
  } catch (err) {
    // Without a valid key we still return businesses, just without project detail.
    return businesses.map((b) => ({ ...b, project: null }));
  }
  const byId = new Map(projects.map((p) => [p.id, p]));
  return businesses.map((b) => ({ ...b, project: b.projectId ? byId.get(b.projectId) || null : null }));
}

// GET /api/businesses — businesses with resolved Linear project detail.
router.get(
  '/',
  asyncHandler(async (req, res) => {
    const businesses = readStore().businesses.filter((business) => inWorkspace(req, business));
    res.json({ businesses: await withProjects(businesses) });
  })
);

// POST /api/businesses — create a business, optionally creating its Linear project.
router.post(
  '/',
  asyncHandler(async (req, res) => {
    const body = req.body || {};
    const name = (body.name ? String(body.name) : '').trim();
    if (!name) return res.status(400).json({ error: 'Business name is required.' });

    const description = body.description ? String(body.description) : '';
    const repository = repositoryFields(body);
    if (repository.error) return res.status(400).json({ error: repository.error });
    let projectId = body.projectId ? String(body.projectId) : null;

    // Optionally create a brand-new Linear project for this business.
    if (!projectId && body.createNewProject) {
      const teamId = body.teamId ? String(body.teamId) : '';
      if (!teamId) {
        return res.status(400).json({ error: 'A team is required to create a new project.' });
      }
      // Auto-attach the configured enrich labels (default ["AI"]) so the new
      // project is immediately picked up by the enrichment scheduler. Gated by the
      // autoLabelNewProjects config toggle.
      const config = getAgentConfig();
      let labelIds = [];
      if (config.autoLabelNewProjects) {
        const labels = await getOrCreateProjectLabels(getApiKey(), config.enrichLabels);
        labelIds = labels.map((l) => l.id);
      }
      const project = await createProject(getApiKey(), {
        name: body.projectName ? String(body.projectName) : name,
        description,
        teamId,
        labelIds,
      });
      projectId = project.id;
    }

    const store = readStore();
    const business = {
      id: slugify(name),
      name,
      description,
      projectId,
      // Provider is stored beside the canonical namespace so a later global
      // connector switch cannot reinterpret this project repository.
      repo: repository.repo,
      repoProvider: repository.repoProvider,
      ...contextFields(req),
      createdAt: new Date().toISOString(),
    };
    if (store.businesses.some((b) => b.id === business.id && inWorkspace(req, b))) {
      return res.status(409).json({ error: 'A business with that name already exists.' });
    }

    const next = { ...store, businesses: [...store.businesses, business] };
    writeStore(next);
    const [enriched] = await withProjects([business]);
    res.status(201).json({ business: enriched });
  })
);

// PUT /api/businesses/:id — update fields / link a Linear project.
router.put(
  '/:id',
  asyncHandler(async (req, res) => {
    const store = readStore();
    const index = store.businesses.findIndex((b) => b.id === req.params.id && inWorkspace(req, b));
    if (index === -1) return res.status(404).json({ error: 'Business not found.' });

    const body = req.body || {};
    const current = store.businesses[index];
    const repository = repositoryFields(body, current);
    if (repository.error) return res.status(400).json({ error: repository.error });
    const updated = {
      ...current,
      name: body.name !== undefined ? String(body.name).trim() || current.name : current.name,
      description: body.description !== undefined ? String(body.description) : current.description,
      projectId:
        body.projectId !== undefined
          ? (body.projectId ? String(body.projectId) : null)
          : current.projectId,
      repo: repository.repo,
      repoProvider: repository.repoProvider,
    };

    const businesses = store.businesses.map((b, i) => (i === index ? updated : b));
    writeStore({ ...store, businesses });
    const [enriched] = await withProjects([updated]);
    res.json({ business: enriched });
  })
);

// DELETE /api/businesses/:id — remove the local business mapping (leaves Linear untouched).
router.delete('/:id', (req, res) => {
  const store = readStore();
  const businesses = store.businesses.filter((b) => b.id !== req.params.id || !inWorkspace(req, b));
  if (businesses.length === store.businesses.length) {
    return res.status(404).json({ error: 'Business not found.' });
  }
  writeStore({ ...store, businesses });
  res.json({ ok: true });
});

module.exports = router;
module.exports.normalizeRepo = normalizeRepo;
module.exports.normalizeRepoProvider = normalizeRepoProvider;
module.exports.repositoryFields = repositoryFields;
module.exports.inWorkspace = inWorkspace;
module.exports.contextFields = contextFields;
